[DCOS-15644] DC/OS Docker on macOS

adamtheturtle opened the pull request:
## High Level Description

Added documentation for running and troubleshooting DC/OS Docker on macOS. It is intended that this documentation is all clearly marked as experimental.

This is a development convenience and I have found it very useful in the last few days.

## Related Issues

https://jira.mesosphere.com/browse/DCOS-15645 - Run DC/OS Docker on macOS without Vagrant

## Testing

### On macOS

I have spun up a cluster in the default configuration and run some integration tests.

I have run `make DOCKER_GRAPHDRIVER=overlay MESOS_SYSTEMD_ENABLE_SUPPORT=true` and seen that this fails as expected:

```
tee: /run/systemd/system/mesos_executors.slice: No such file or directory
[Unit]
Description=Mesos Executors Slice
sudo: systemctl: command not found
make: *** [Makefile:133: /run/systemd/system/mesos_executors.slice] Error 1
```

The following command was run to see integration tests passing:

```
cd /opt/mesosphere/active/dcos-integration-test/util
DCOS_PYTEST_CMD='pytest -s -x test_iam.py' DCOS_NUM_MASTERS='1' DCOS_NUM_AGENTS='2' DCOS_LOGIN_UNAME='admin' DCOS_LOGIN_PW='admin' TEST_DCOS_RESILIENCY='false' SLAVE_HOSTS='172.17.0.3' PUBLIC_SLAVE_HOSTS='172.17.0.4' /bin/bash run_integration_test.sh
```

adamtheturtle commented:
Closing while I work on this.

adamtheturtle commented:
Ping for review @mhrabovcin @karlkfi @tillt

adamtheturtle commented on Makefile:
```diff
@@ -287,6 +299,8 @@ docker run -dt --privileged \
 	--add-host "$(REGISTRY_HOST):$(shell $(IP_CMD) $(MASTER_CTR)1 2>/dev/null || echo 127.0.0.1)" \
 	$(DOCKER_IMAGE);
 sleep 2;
+docker exec $(1)$(2) mkdir -p /var/lib/dcos
```
TODO: Comment where this path comes from, what it does

adamtheturtle replied:
Summary: This is one of the common files in https://github.com/dcos/dcos/blob/master/packages/mesos/extra/dcos-mesos-slave.service. We disable the systemd of Mesos if we are not using systemd on the host because that requires systemd on the host.

We use a "common" file so that we only write to one file for agent and public agent nodes.

adamtheturtle replied:
Now done above

karlkfi replied:
Why are we doing multiple docker execs instead of a single one with a multi-line bash script?
Why are we doing docker execs at all instead of just making the docker run a bash script?

adamtheturtle replied:
> Why are we doing multiple docker execs instead of a single one with a multi-line bash script?

No reason - though I don't see why it is better - performance wasn't really my concern if that is yours. I am happy to change this though.

> Why are we doing docker execs at all instead of just making the docker run a bash script?

I continued the path of using execs here but I can change that if you like. Again, I have no strong preference.

Please let me know how to proceed and I will.

karlkfi replied:
I see you were just following the existing pattern. Tho when you added a `/bin/bash -c` line it made the pattern look kinda silly. Plus, when you use `/bin/bash -c` you lose the error checking. You're probably want to do `/bin/bash -ce` or go multi-line and use `set -o errexit -o nounset -o pipefail`.

adamtheturtle replied:
I will do the latter and I will implement that.

joerg84 commented:
@karlkfi

karlkfi commented on Makefile:
```diff
@@ -119,17 +127,19 @@ postflight: ## Polls DC/OS until it is healthy (5m timeout)
 
 master: ## Starts the containers for DC/OS masters.
-	$(foreach NUM,$(shell seq 1 $(MASTERS)),$(call start_dcos_container,$(MASTER_CTR),$(NUM),$(MASTER_MOUNTS) $(TMPFS_MOUNTS) $(CERT_MOUNTS) $(HOME_MOUNTS) $(VOLUME_MOUNTS)))
+	$(foreach NUM,$(shell seq 1 1 $(MASTERS)),$(call start_dcos_container,$(MASTER_CTR),$(NUM),$(MASTER_MOUNTS) $(TMPFS_MOUNTS) $(CERT_MOUNTS) $(HOME_MOUNTS) $(VOLUME_MOUNTS)))
```
what does adding this `1` do?

adamtheturtle replied:
Please see "seq changes" in the PR description.

karlkfi replied:
👍

karlkfi replied:
Turns out this errors if you set AGENTS or PUBLIC_AGENTS to zero.
```
$ make MASTERS=1 AGENTS=0 PUBLIC_AGENTS=0
...
seq: needs negative decrement
```

adamtheturtle replied:
I believe it still works and creates that cluster - right? I can't test now. Maybe I just need to hide that.

karlkfi replied:
This doesn't seem to be a new problem. The rest of the code doesn't allow zero agents.

adamtheturtle replied:
The problem before was if you give AGENTS=0 as an argument you got 2 agents. Now you get 0 I believe.

karlkfi replied:
Here's a workaround:
```
$(foreach NUM,$(shell [[ $(MASTERS) == 0 ]] || seq 1 1 $(MASTERS)),$(call run_dcos_install_in_container,$(MASTER_CTR),$(NUM),master))
```
For the master/agent/public_agent targets you also have to add an echo or something so the target isn't just empty:
```
@echo "+ Starting master nodes"
```

adamtheturtle replied:
What is this working around? The error output or a symptom which affects use?

karlkfi replied:
```
$ seq 1 1 0; echo $?
seq: needs negative decrement
1
```

non-zero exit codes cause make to exit. So if you set to zero it will exit before finishing.

karlkfi commented on distros/generate.sh:
```diff
@@ -9,7 +9,13 @@ set -e
-cd "$(dirname "$(readlink -f "$BASH_SOURCE")")"
+# `readlink -f` does not work on macOS.
+# To get `greadlink`, run `brew install coreutils`.
+if [ "$(uname)" == "Darwin" ]; then
```
Can we not do this approach with gnu sed?

adamtheturtle replied:
Do you mean `gnu seq`? If so, yes. Do you prefer that? I preferred the other because it was cross-platform. I again don't have a strong preference, so I will follow your preference.

adamtheturtle replied:
As for GNU sed, `sed` is used in the DC/OS repo and I preferred to not make changes there for this purpose.

adamtheturtle replied:
Remember that this requires an install while changing `seq 1` to `seq 1 1` does not.

karlkfi replied:
I meant GNU sed, but it sounds like you're implying that's upstream in dcos/dcos, not in dcos-docker. However, I was under the impression that the installer didn't work on Mac. Are you saying that GNU sed was the only blocker for that?

adamtheturtle replied:
https://github.com/dcos/dcos/blob/b5f75d38e85e26ead42a1c41ec70c7c09591c6c0/gen/build_deploy/bash/dcos_generate_config.sh.in#L19 is the line that caused an issue. With this PR and the instructions in it, we can run DC/OS Docker on macOS if that answers your question.

adamtheturtle replied:
> we can run DC/OS Docker on macOS

See "Testing" in the PR description for a more in-depth explanation of what I mean by this.

karlkfi replied:
If `sed '0,/^#EOF#$/d' $0 | tar xv` is the only line blocking the installer from working on Mac without installing GNU sed, can we just change that line to be more cross platform?

karlkfi replied:
Alternatively, can we modify it to support `brew install gnu-sed` without `--with-default-names`? The default names thing has a lot of side effects on anything else running on your mac. I'm not comfortable recommending it.

adamtheturtle replied:
Possibly - however is that better than having the workaround here? I would prefer not to make a change to DC/OS for this purpose but I can do if you request it.

karlkfi replied:
If it's as trivial a change as  you say, I'd much rather have the change upstreamed. Then I can also trivially apply this method to dcos-vagrant to build on the host instead of using a bootstrap VM.

adamtheturtle replied:
How about I change the docs to make the dangers clear and provide two alternatives:
* Uninstall straight after,
* Temporarily create an alias / symlink
?

> If it's as trivial a change as you say

How about I make a PR and see what the reaction is. I would like this review to continue as if the PR wouldn't land, and get this merged if we are waiting on that DC/OS PR. If the DC/OS PR does land then I can take the `brew` install out of the README. In the meantime I can put a warning in the README here.

karlkfi replied:
Here's a better solution that using readlink:
```
cd "$(cd "$(dirname "${BASH_SOURCE}")" && pwd -P)"
```

karlkfi replied:
Actually, that's maybe not technically the same, but I wonder what platform uses a symlink for BASH_SOURCE...

karlkfi replied:
Ok, so it turns out the existing readlink option only supports one layer of indirection via symlink, and only if the generate.sh script itself is an absolute (not relative) symlink.

Since we're not symlinking the generate.sh script, I think we can completely do away with the readlink and go with something simpler, like 
```
cd "$(dirname "${BASH_SOURCE[0]}")"
```

adamtheturtle commented:
After discussion with @karlkfi and my team I will do the following:

* Split out the systemd work from the macOS work and submit the former for approval, making the changes which Karl mentioned here
* Get to the macOS work when possible and prioritised - this was started as a side project and the review requirements put perhaps make too much work necessary to complete this at the moment

In particular the controversial change with macOS relates to `sed`. `sed` is used at https://github.com/dcos/dcos/blob/b5f75d38e85e26ead42a1c41ec70c7c09591c6c0/gen/build_deploy/bash/dcos_generate_config.sh.in#L19 in DC/OS. DC/OS provides no support for macOS. This proposed change overwrites macOS's `sed` with GNU sed which has side-effects. The review request was that I made a patch to DC/OS which made that `sed` command platform agnostic, and added a test for regression safety of that change on macOS. That test might involve checking that DC/OS Docker could spin up successfully on macOS.

A compromise was reached whereby the following plan was agreed upon:
* I will submit a patch to DC/OS with macOS support without tests, noting my preference to not document or commit to even experimental support for macOS in DC/OS
* Feedback from the team will determine whether the best approach is:
    - Untested change to DC/OS
    - Tested change to DC/OS (this would require scheduling from my team if I were to do it)
    - Change to DC/OS Docker as in this branch
    - No change

@karlkfi ^ I hope that represents the conversation we had accurately.

adamtheturtle commented:
For posterity, it is possible to fix this by changing the `sed` command to ` awk '/^#EOF#/ { f = 1 } f' $0 | tail -n +2 | tar xv` and running `brew install gawk` on macOS. However, I believe that this has similar issues to `brew install gnu-sed --with-default-names`. Please confirm @karlkfi that this is not a suitable solution.

karlkfi commented:
Looks like Docker 1.13+ will be required for this to work on Mac with docker-machine, because 1.12 has a bug that validates the tmpfs options on the client, which causes a `docker: Invalid tmpfs option ["rw"].` error. 

Bug: https://github.com/moby/moby/issues/22205
Fix: https://github.com/moby/moby/pull/28440

Solution: noop (req already say Docker 1.13.1)

For dcos-vagrant I've been keeping a troubleshooting file for things like this that have a known workaround. It helps SEO and people can resolve their own issues even if they forgot to read the requirements.

karlkfi commented on README.md:
```diff
@@ -223,6 +255,12 @@ Most standard distribution kernels should have this by default. On some
 older kernels you may need to manually install this module with
 `modprobe dummy` before starting the container cluster.
 
+# Docker for Mac out of space
```
this probably applies to all dockers...

karlkfi commented on README.md:
```diff
@@ -20,10 +23,16 @@ DC/OS Docker is designed to optimize developer cycle time. For a more production
 - git
 
-### Mac
+### macOS (Experimental)
```
prefer just `### Mac` with a bold line after saying it's experimental and why. That way the page anchor link doesn't change later.

karlkfi commented on README.md:
```diff
@@ -20,10 +23,16 @@ DC/OS Docker is designed to optimize developer cycle time. For a more production
+  See "Graphdriver/Storage driver"
+- git
+
+### Vagrant on either platform
```
just `### Vagrant` for a clean anchor link, then explain platform support on next line.

adamtheturtle replied:
I will but yes this is another reason why `ReST` is so much better than MD for these purposes - you choose the anchor.

karlkfi commented on README.md:
```diff
@@ -40,10 +49,8 @@ DC/OS Docker is designed to optimize developer cycle time. For a more production
 1. Move the installer to `dcos_generate_config.sh` in the root of this repo directory.
 
-**The following steps are REQUIRED on Macs and OPTIONAL on Linux.**
+**The following steps required if using Vagrant on either platform.**
```
- `on either platform`

karlkfi commented on README.md:
````diff
@@ -98,9 +114,15 @@ On **Mac OS X**:
 ```
 
-Once routing is set up, you can access DC/OS directly from the host.
+### Set up routing (macOS host)
````
kinda awkward to have this between the setup and cleanup for vagrant...

karlkfi commented:
I'd merge and tweak it myself, but 41... commits! can you squash it?

adamtheturtle commented:
> I'd merge and tweak it myself, but 41... commits! can you squash it?

I will do because it is the standard here but really I don't see the problem with an honest, large history - the `--first-parent` log of PR merges gives a "clean" history. Crafting history into "How I wish this had been created" is something I haven't yet found any value in.

karlkfi commented:
I don't always ask for squashing. This and the sed case are exceptional because the PR is radically different than when it started. We're using PRs to have conversations and experiment, and I don't think all that back and forth needs to go in the commit log. I want to be able to look in the blame log and see why it is the way it is, not excavate an archaeological dig where the last commit say "because a reviewer asked me to"

karlkfi commented:
More succinctly: the history of master should represent shared history. 

If a branch changes code from X to Y to Z before being merged, then the only thing that matters to users of the log after merge is why Z is better than X. So we optimize for that use case.

karlkfi commented:
Thanks for working through this, @adamtheturtle !
